### DISCUSSION TITLE
How to use retry/2

### DISCUSSION THREAD

**Discussion by aidalgol** [keathley/wallaby#608] [Q&A] [answered]
I have the following in one of my Wallaby tests:
```elixir
   retry(fn ->
      if current_path(session) == "/" do
        IO.puts("At root")

        session
        |> assert_has(Query.css(".notification.is-info"))
       else
        {:error, :url_not_refreshed}
      end
    end)
 ```

And the test passes, even though the success path is never hit, because the message `At root` is never printed.  This is in a test after an action that should redirect to the root path, hence the `retry` instead of just asserting immediately after the navigation action.  What I was expecting to happen was for `retry/2` to cause my test to fail if the sync result is only ever `{:error, any()}` within some timeout, but that is clearly not happening.  How can I achieve this?

**Comment by mhanberg** [comment, marked as answer]
Can you elaborate a little more on the test scenario?

I am curious as to why you are resorting to using the retry function.

But, the reason why you're test is passing right now is because `retry/1` expects the callback to return an ok/error tuple, and `retry/1` will then pass that value through, but you aren't asserting on anything. The following would fail your test if the callback passed to `retry/1` does not return `{:ok, <some value>}` within the timeout.

```elixir
assert {:ok, _session} =
   retry(fn ->
      if current_path(session) == "/" do
        IO.puts("At root")

        {:ok, assert_has(session, Query.css(".notification.is-info"))}
       else
        {:error, :url_not_refreshed}
      end
    end)
```

**Reply by aidalgol** [reply]
OK, thanks.  That makes much more sense.

As for why I am using `retry/2` in the first place is that the preceding part of the test fills out the login form and clicks the submit button, which should then return a 302 response, and then redirect to `/`, so the retry is in there to wait for the redirect to complete.

**Reply by mhanberg** [reply]
> As for why I am using retry/2 in the first place is that the preceding part of the test fills out the login form and clicks the submit button, which should then return a 302 response, and then redirect to /, so the retry is in there to wait for the redirect to complete.

Ahh, I see. `assert_has/2` internally calls retry, so you shouldn't need to call it yourself.

**Reply by aidalgol** [reply]
Ah, good to know.  I'm guessing `current_path/2` does not, though?

**Reply by mhanberg** [reply]
There is nothing to block on for `current_path/2`. It asks the browser for the current path and returns the result.

You shouldn't need to wait for the browser to be on the right page before writing your assertion.

**Reply by aidalgol** [reply]
Oh, of course.  Duh.  OK, I think that's everything then.  Thanks for your help!
